recipes/jenkins.rb

```diff
@@ -31,6 +31,53 @@
 apt_update
 
 package 'openjdk-8-jdk-headless'
+# Jenkins downgrade protection
+#
+# The Jenkins package has transitioned to using systemd units instead of
+# sysvinit style init scripts. In order to maintain this cookbook the last
+# version of Jenkins which is packaged with a sysvinit script is pinned.
+# However Jenkins cannot be safely downgraded and some configurations of
+# Jenkins will continue to function once initially configured even after the
+# systemd switch.
+# To protect users from unintentional downgrades we're going to do something really ugly here.
+package 'jenkins' do
+  action :lock
+  only_if { -> do
+    jenkins_info = `dpkg -s jenkins`
+    if $?.exitstatus != 0
+       return false
+    end
+
+    version_line = jenkins_info.lines.select{|line| line =~ /^Version: /}.first
+    if version_line.nil?
+      return false
+    end
+
+    # Transform "Version: 2.319.1\n" to ["2", "319", "1"]
+    version_components = version_line.chomp.split(": ")[1].split(".")
+    # The first systemd versions are 2.335 (weekly) and 2.332.1 (LTS)
+    if version_components[0].to_i > 2 or version_components[1].to_i >= 332
```

Comment:
Could this be cleaned up with the "spaceship operator"?

Reply:
`<=>` implements three-way comparison for a class. I don't actually think it helps us unless we parse these version components into an actual class and implement the component level comparisons. Although you could be seeing something tidy that I'm not.

The logic here is complex enough that I was optimizing for plain readability rather than conciseness. I suppose that if we had a Comparable version class here we would be able to write

```
version_installed = Version(version_line.chomp.split(": ")[1].split("."))
if version >= Version([2,332,1])
```